src/main/java/org/apache/commons/io/IOUtils.java

```diff
@@ -1147,31 +1201,43 @@ public static boolean contentEqualsIgnoreEOL(final Reader input1, final Reader i
                             }
                             continue;
                         default:
-                            if (lastState1 == LastState.r) {
-                                lastState1 = LastState.newLine;
+                            // if input2's next is normal.
+                            //  if input1's last is '\r', then it can become "\r\n", then legal.
+                            //  otherwise illegal.
+                            if (lastState1 == LastState.R) {
+                                lastState1 = LastState.NEW_LINE;
                                 nowCheck1++;
                                 continue;
                             } else {
                                 return false;
                             }
                     }
                 default:
+                    // if input1's next is normal.
                     switch (charArray2[nowCheck2]) {
                         case '\n':
-                            if (lastState2 == LastState.r) {
-                                lastState2 = LastState.newLine;
+                            // if input2's next is '\n'.
+                            //  if input2's last is '\r', then it can become "\r\n", then legal.
+                            //  otherwise illegal.
+                            if (lastState2 == LastState.R) {
+                                lastState2 = LastState.NEW_LINE;
                                 nowCheck2++;
                                 continue;
                             } else {
                                 return false;
                             }
                         case '\r':
+                            // if input2's next is '\r'.
+                            // illegal.
                             return false;
                         default:
+                            // if input2's next is normal.
+                            //  if equal then legal.
+                            //  otherwise illegal.
                             if (charArray1[nowCheck1] != charArray2[nowCheck2]) {
                                 return false;
                             }
-                            lastState1 = lastState2 = LastState.normal;
+                            lastState1 = lastState2 = LastState.NORMAL;
                             nowCheck1++;
                             nowCheck2++;
                             continue;
```

Comment:
Most of the comments above are unnecessary if the variables are well named and commented.
What is missing is how the algorithm works, and what the various combinations of state actually mean in terms of the algorithm.

I'm not clear why the code sometimes checks for '\r' and sometimes uses the enum.
Why not use a variable containing the last character?

==

I suspect the code could be much simplified by using a suitable filter on the inputs.
There are several examples in IO, and NET has FromNetASCIIInputStream which deals with CRLF conversions

Reply:
> Most of the comments above are unnecessary if the variables are well named and commented.
> What is missing is how the algorithm works, and what the various combinations of state actually mean in terms of the algorithm.
> 
> I'm not clear why the code sometimes checks for '\r' and sometimes uses the enum.
> Why not use a variable containing the last character?
> 
> ==
> 
> I suspect the code could be much simplified by using a suitable filter on the inputs.
> There are several examples in IO, and NET has FromNetASCIIInputStream which deals with CRLF conversions



> It occurs to me that the code is effectively buffering the output from a BufferedReader (or BufferedInputStream).
> One would expect these classes to be reasonably fast, however the JVM has to do locking and other checks in order to support multi-threading. It has to do this for each read() call.
> 
> Rather than implement the buffering directly in the compare methods, it might be better to implement a generic, non-threadsafe filter that can be used in situations such as these.
> 
> The original code should then work without any change other than to add the filter.



> It occurs to me that the code is effectively buffering the output from a BufferedReader (or BufferedInputStream).
> One would expect these classes to be reasonably fast, however the JVM has to do locking and other checks in order to support multi-threading. It has to do this for each read() call.
> 
> Rather than implement the buffering directly in the compare methods, it might be better to implement a generic, non-threadsafe filter that can be used in situations such as these.
> 
> The original code should then work without any change other than to add the filter.

Hi.
I used the filter idea you mentioned and re-write another implementation, named contentEqualsIgnoreEOLNew2.
You can see it in the latest commit.
A fast performance test shows that it is SLOWER than my giant function (called contentEqualsIgnoreEOLNew1), but still very much FASTER than the original implementation in commons-io.

Reply:
@sebbASF 
> Most of the comments above are unnecessary if the variables are well named and commented.
> What is missing is how the algorithm works, and what the various combinations of state actually mean in terms of the algorithm.

I added more comments, please have a look.

> I'm not clear why the code sometimes checks for '\r' and sometimes uses the enum.
> Why not use a variable containing the last character?

because in some cases even if last character is '\r', we should treated it as a new line start.
for example, in
"aaaaa\r\naaaaa" and "aaaaa\raaaaa",
although in second String it is \r, it still need to treat it as new line, so new line here not mean LF, but "new line".

though, we can use an int instead of enum.
I tried, and that can work, and even be faster than the enum version....

> I suspect the code could be much simplified by using a suitable filter on the inputs.

Yes I think so. But it will be a performance cost in doing so.